File: tools/tal_support/as274_fc/C/t1.c
#include <stdio.h>
#include "as274c.h"
#include "misc.h"

int main()
{
    double **x, *y, *d, *rbar, *thetab, *xrow, sserr, *beta;
    double *tol, *rss;
    int i, j, *vorder, *lindep, error;

    x = dmatrix(0, 10, 0, 4);
    y = dvector(0, 10);
    d = dvector(0, 4);
    rbar = dvector(0, 9);
    thetab = dvector(0, 4);
    xrow = dvector(0, 4);
    beta = dvector(0, 4);
    tol = dvector(0, 4);
    rss = dvector(0, 4);
    vorder = ivector(0, 4);
    lindep = ivector(0, 4);

    y[0] = 65647.;
    y[1] =  70638.;
    y[2] =  75889.;
    y[3] =  81399.;
    y[4] =  87169.;
    y[5] =  93202.;
    y[6] =  99503.;
    y[7] = 106079.;
    y[8] =  112939.;
    y[9] =  120094.;
    y[10] =  127557.;
    for (i=0; i<5; i++) vorder[i] = i;

    printf("Fitting nasty cubic; 1st 4 betas should be 1\n\n");
    for (i=0; i<11; i++) {
        double tmp = i+40;
        x[i][0] = tmp;
        x[i][1] = tmp*tmp;
        x[i][2] = x[i][0] * x[i][1];
        x[i][3] = x[i][0] + x[i][1];
    }
    error = clear(5, 10, d, rbar, thetab, &sserr);
    if (error) printf("clear() returned %d\n", error);

    for (i=0; i<11; i++) {
        xrow[0] = 1;
        for (j=0; j<4; j++) xrow[j+1] = x[i][j];
        error = includ(5, 10, 1.0, xrow, y[i], d, rbar, thetab, &sserr);
        if (error) printf("includ() returned %d at row %d\n",
                          error, i);
    }

    error = tolset(5, 10, d, rbar, tol);
    if (error) printf("tolset() returned %d\n", error);
    error = sing(5, 10, d, rbar, thetab, &sserr, tol, lindep);
    if (error) printf("sing() returned %d\n", error);
    printf("SSerr = %17g, expect 286.\n", sserr);

    for (i=0; i<6; i++) {
        error = regcf(5, 10, d, rbar, thetab, tol, beta, 5);
        if (error) printf("regcf() returned %d\n", error);
        printf("\n\nVariable order: ");
        for (j=0; j<5; j++) printf("%d ", vorder[j]); printf("\n");
        putdvec("Beta", beta, 0, 4);
        putdvec("d", d, 0, 4);
        printf("rbar matrix:\n"); pr_utdm_v(rbar, 4, 14, 6);
        putdvec("thetab", thetab, 0, 4);

        /* rotate through 1 */
        error = ss(5, d, thetab, &sserr, rss);
        if (error) printf("ss() returned %d\n", error);
        error = vmove(5, 10, vorder, d, rbar, thetab, rss, 1, 5, tol);
        if (error) printf("vmove() returned %d\n", error);
    }

    return 0;
}

File: tools/tal_support/as274_fc/F/t1.f
      program test1
c
c     Test vmove.
c     The LS fit to the data is:  Y = 1 + X + X**2 + X**3
c     i.e. all of the regression coefficients should be exactly 1.0.
c     An extra variable equal to X + X**2 is inserted to give a singularity.
c
      double precision x(11,4), y(11), d(5), rbar(10), thetab(5)
	  double precision xrow(5), yelem,
     +   sserr, beta(5), work(5), tol(5), rss(5)
      integer i, j, ifault, vorder(5)
      logical lindep(5)

      data y/65647., 70638., 75889., 81399., 87169., 93202., 99503.,
     +   106079., 112939., 120094., 127557./
      data vorder/0, 1, 2, 3, 4/

      write(*, *)'Fitting nasty cubic'
      write(*, *)'1st 4 regression coefficients should equal 1.0'
      write(*, *)

      do 10 i = 1, 11
	x(i,1) = i + 39
	x(i,2) = x(i,1)**2
	x(i,3) = x(i,1) * x(i,2)
	x(i,4) = x(i,1) + x(i,2)
   10 continue
c
c     Use includ to form the orthogonal reduction.
c
      do 20 i = 1, 5
	d(i) = 0.0
	thetab(i) = 0.0
   20 continue
      do 30 i = 1, 10
   30 rbar(i) = 0.0
      sserr = 0.0

      do 50 i = 1, 11
	xrow(1) = 1.0
	do 40 j = 1, 4
   40   xrow(j+1) = x(i,j)
	yelem = y(i)
	call includ(5, 10, 1.0d0, xrow, yelem, d, rbar, thetab, sserr,
     +      ifault)
   50 continue
c
c     Call tolset to set tolerances, then sing to detect singularities.
c
      call tolset(5, 10, d, rbar, tol, work, ifault)
      call sing(5, 10, d, rbar, thetab, sserr, tol, lindep, work,
     +    ifault)
      write(*, *)'sserr = ', sserr, '   Should be 286.000'
c
c     Calculate regression coefficients, using vmove to cycle through
c     the order of the variables.
c
      do 60 i = 1, 6
	call regcf(5, 10, d, rbar, thetab, tol, beta, 5, ifault)
	write(*, *)'Variable order:', vorder
	write(*, 900) beta
  900   format(' Regn. coeffs.:'/ 1x, 5g15.7)
	write(*, 910) d, rbar, thetab
  910   format(' d: ', 5g15.7/ ' rbar:'/1x,4g15.7/16x,3g15.7/31x,2g15.7/
     +      46x,g15.7/ ' thetab: '/1x, 5g15.7//)
c
c     Calculate residual sums of squares (RSS).
c
	call ss(5, d, thetab, sserr, rss, ifault)
c
c     Move variable in 1st position to the last.
c
	call vmove(5, 10, vorder, d, rbar, thetab, rss, 1, 5, tol,
     +          ifault)
   60 continue
      end

